torch/csrc/jit/codegen/cuda/lower_interleaved_loop.cpp

```diff
@@ -0,0 +1,712 @@
+      main_loop, IdMappingMode::LOOP);
+  auto concrete_sub_loop = GpuLower::current()->caMap()->getConcreteMappedID(
+      sub_loop, IdMappingMode::LOOP);
+
+  // Insert sub loops from this tv
+  auto main_loop_entry_it =
+      concrete_main_loop_to_subloop_map_.find(concrete_main_loop);
+  TORCH_INTERNAL_ASSERT(
+      main_loop_entry_it != concrete_main_loop_to_subloop_map_.end(),
+      "unknown main loop ",
+      main_loop->toString());
+  main_loop_entry_it->second.pushBack(concrete_sub_loop);
+
+  // Insert interleaved tvs.
+  auto tv_entry_it =
+      concrete_main_loop_to_interleaved_tv_.find(concrete_main_loop);
+  TORCH_INTERNAL_ASSERT(
+      tv_entry_it != concrete_main_loop_to_interleaved_tv_.end(),
+      "unknown main loop");
+  tv_entry_it->second.pushBack(tv);
+}
+
+void InterleaveLoopInfo::collectInterleavedSubLoops() {
+  for (auto tv : used_tvs_) {
+    IterDomain* main_loop = nullptr;
+    for (auto leaf_id : tv->domain()->domain()) {
+      if (isMainLoop(leaf_id)) {
+        TORCH_INTERNAL_ASSERT(
+            main_loop == nullptr,
+            tv,
+            "has nested main loop ",
+            main_loop->toString(),
+            " and ",
+            leaf_id->toString(),
+            " which is not yet supported");
+
+        auto maybe_subloop = getMaybeSubloop(tv, leaf_id);
+        TORCH_INTERNAL_ASSERT(
+            maybe_subloop.has_value(),
+            tv->toString(),
+            " cannot be interleaved within ",
+            leaf_id);
+        insertEntry(tv, leaf_id, maybe_subloop.value());
+        main_loop = leaf_id;
+      }
+    }
+  }
+}
+
+// Validation of double buffering topology of interleaved expressions:
+//  see [Supported Interleaving Cases]
+void InterleaveLoopInfo::validate() {
+  // Validate expression consistency after interleaving
+  for (auto& main_loop_entry : concrete_main_loop_to_interleaved_tv_) {
+    validateMainLoop(main_loop_entry.first, main_loop_entry.second);
+  }
+}
+
+// Returns true if the given tv is an "exit tv",
+//  see [Supported Interleaving Cases].
+bool InterleaveLoopInfo::isExitTv(
+    TensorView* tv,
+    const TensorViewVector& interleaved_tvs) {
+  // Output is always an exit
+  if (tv->isFusionOutput()) {
+    return true;
+  }
+
+  for (auto use : fusion_->unordered_uses(tv)) {
+    // Check if any immediate consumer of tv is interleaved.
+    for (auto consumer_tv :
+         ir_utils::filterByType<TensorView>(use->outputs())) {
+      if (interleaved_tvs.has(consumer_tv)) {
+        return false;
+      }
+    }
+  }
+
+  // No immediate consumer of tv is interleaved so the tv is an exit tv.
+  return true;
+}
+
+void InterleaveLoopInfo::validateMainLoop(
+    IterDomain* concrete_main_loop,
+    const TensorViewVector& interleaved_tvs) {
+  // [Supported Interleaving Cases]
+  // All the expressions that are inside the main loop or subloop can
+  //  only be 3 cases:
+  // 1. It's double/circular buffered across a loop that's either at or on the
```

Comment:
Can't follow this case. 

What I can't yet figure out is what the underlying generic condition this transformation must satisfy. Generally speaking, it seems safe if there's no data dependency between the subloop TVs, which basically corresponds to the Point 3. In the case of Point 2, it is also safe despite the data dependency because the dependency is constrained inside the sub loop, right? I can't wrap my head around the Point 1 yet, though.